jax/_src/lax/lax.py

```diff
@@ -2991,8 +2994,34 @@ def _concatenate_batch_rule(batched_args, batch_dims, *, dimension):
 ad.primitive_transposes[concatenate_p] = _concatenate_transpose_rule
 batching.primitive_batchers[concatenate_p] = _concatenate_batch_rule
 
-def _concatenate_lower(ctx, *xs, dimension):
-  return mhlo.ConcatenateOp(xs, mlir.i64_attr(dimension)).results
+def _concatenate_staging_rule(
+    trace, *args, dimension, shape):
+  dyn = args[- _nr_dynamic_dimensions(shape):]
+  params = dict(dimension=dimension, shape=shape)
+  if not dyn:
+    return trace.default_process_primitive(concatenate_p, *args, params)
+  aval = core.DShapedArray(_merge_dyn_shape(shape, dyn), args[0].dtype, args[0].weak_type)
+  return _dyn_shape_staging_rule(trace, concatenate_p, aval, *args, **params)
+pe.custom_staging_rules[concatenate_p] = _concatenate_staging_rule
+
+def _concatenate_typecheck_rule(*args, dimension, shape):
+  nr_dynamic_dimensions = _nr_dynamic_dimensions(shape)
+  if not nr_dynamic_dimensions:
+    out_aval, effects = concatenate_p.abstract_eval(
+        dimension=dimension, shape=shape)
+    return [out_aval], effects
+  else:
+    dyn_shape = args[- nr_dynamic_dimensions:]
+    out_shape = _merge_dyn_shape(shape, dyn_shape)
+    out_shape = [x.val if type(x) is core.Literal else x for x in out_shape]
+    out_aval = core.DShapedArray(tuple(out_shape), args[0].aval.dtype, args[0].aval.weak_type)
```

Comment:
I do not understand the different form of DShapedArray in the staging rule and the type checking rule. I understand that in the first case we have Tracers and in the latter we don't, but isn't it confusing that DShapedArray has two forms?

Reply:
Does [this comment](https://github.com/google/jax/blob/2c72858928440be9d272fe5d04656cf1dcac20b5/jax/core.py#L1409) help explain?

The staging rule needs to put Tracers in shape tuples, while the typechecking rule needs to put Vars in shape tuples.